source/extensions/tracers/datadog/event_scheduler.cc

```diff
@@ -0,0 +1,83 @@
+#include "source/extensions/tracers/datadog/event_scheduler.h"
+
+#include <memory>
+#include <utility>
+
+#include "source/common/common/assert.h"
+
+#include "datadog/json.hpp"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+
+EventScheduler::EventScheduler(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+EventScheduler::~EventScheduler() {
+  for (const auto& timer : timers_) {
+    timer->disableTimer();
+  }
+}
+
+EventScheduler::Cancel
+EventScheduler::schedule_recurring_event(std::chrono::steady_clock::duration interval,
+                                         std::function<void()> callback) {
+  auto self = std::make_shared<Event::Timer*>();
+  // Yes, a shared pointer to a pointer.
+  //
+  // Both the timer callback (argument to `createTimer`, below) and the returned
+  // cancellation function need a handle to the `Event::Timer`. The timer
+  // callback needs it so that it can reschedule the next round when the timer
+  // fires (that's how this is a _recurring_ event). The cancellation function
+  // needs it so that it can call `disableTimer` and remove the timer from
+  // `timers_`.
+  //
+  // Since we don't have a handle to the `Event::Timer` until `createTimer`
+  // returns, we need a "box" out of which the timer callback can extract the
+  // created timer. We then put the `Event::Timer*` in the "box" after
+  // `createTimer` returns the timer. `self` is the "box."
+  //
+  // The cancellation function returned by this function refers to the
+  // `Event::Timer` via a raw pointer (as does the timer callback, indirectly).
+  // The actual lifetime of the pointed-to `Event::Timer` is determined by its
+  // presence in `timers_`. The `Event::TimerPtr` returned by `createTimer` is
+  // moved into `timers_`.
+
+  Event::TimerPtr timer = dispatcher_.createTimer([self, interval, callback = std::move(callback)] {
+    (**self).enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(interval));
+    callback();
+  });
+
+  Event::Timer* timer_raw = timer.get();
+  *self = timer_raw;
+
+  timers_.emplace_back(std::move(timer));
+
+  timer_raw->enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(interval));
+
+  return [this, timer = timer_raw]() mutable {
+    if (!timer) {
+      return; // idempotent
+    }
+
+    timer->disableTimer();
+    auto found = std::find_if(timers_.begin(), timers_.end(),
```

Comment:
does this need to be fast? Should we use an associative structure?

Reply:
Vector is fine for the reason discussed in comments on the header file. I'll add a code comment.

Reply:
I changed it to `absl::flat_hash_set`. Initially I avoided it because I thought that heterogeneous lookups wouldn't be supported without some boilerplate, but it turns out they work out of the box.

A flat hash set has very little overhead compared to a small vector, so we're good.